include/oneapi/dpl/pstl/hetero/dpcpp/utils_ranges_sycl.h

```diff
@@ -309,10 +309,60 @@ __require_access_range(sycl::handler& __cgh, oneapi::dpl::__internal::tuple<_Ran
                                   ::std::make_index_sequence<__num_ranges>());
 }
 
+template <typename...>
+constexpr auto
+__contains_host_pointer_probe(...) -> std::false_type;
+
+// for std::ranges::ref_view
+#if _ONEDPL_CPP20_RANGES_PRESENT
+template <std::ranges::range Rng>
+constexpr auto
+__contains_host_pointer_probe(std::ranges::ref_view<Rng>*) -> std::true_type;
+#endif // _ONEDPL_CPP20_RANGES_PRESENT
+
+template <typename T>
+using __remove_cv_ref_t = typename std::remove_cv<typename std::remove_reference<T>::type>::type;
+
+template <typename T>
+struct __contains_host_pointer : decltype(__contains_host_pointer_probe(static_cast<T*>(nullptr))){};
+
+template <typename...>
+struct __contains_host_pointer_on_any_layers;
+
+// 1. Checking the top-level view
+// 2. Searching view with host pointer in internal views(layers)
+template <typename _View>
+struct __contains_host_pointer_on_any_layers<_View>
+    : std::disjunction<
+          __contains_host_pointer<__remove_cv_ref_t<_View>>,
+          std::conditional_t<oneapi::dpl::__ranges::pipeline_base_range<_View>::has_next_layer_t::value,
+                             __contains_host_pointer_on_any_layers<
+                                 typename oneapi::dpl::__ranges::pipeline_base_range<_View>::next_layer_view_t>,
+                             std::false_type>>
+{
+};
+
+// 1.1. Checking the top-level view
+// 1.2. Searching view with host pointer in internal views(layers)
+// 2.1. Searching view with host pointer in the rest of views
+template <typename _View, typename... _Views>
+struct __contains_host_pointer_on_any_layers<_View, _Views...>
+    : std::disjunction<__contains_host_pointer_on_any_layers<_View>, __contains_host_pointer_on_any_layers<_Views...>>
+{
+};
```

Comment:
I do not see this specialization used anywhere, except the test. If it is not needed for the implementation, let's instead specialize `contains_host_pointer_on_any_layers_v` appropriately.

Reply:
Fixed: this trait and extra test functional ha been removed.